File: courses/lesson5_cars/src/main/java/by/academy/lesson5/cars/WorkListActivity.java
package by.academy.lesson5.cars;

import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.text.Editable;
import android.view.View;
import android.widget.EditText;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.floatingactionbutton.FloatingActionButton;

import by.academy.lesson5.cars.data.CarInfoEntity;
import by.academy.lesson5.cars.data.DatabaseInfo;
import by.academy.lesson5.cars.data.WorkInfoDAO;
import by.academy.lesson5.cars.data.WorkInfoEntity;

public class WorkListActivity extends AppCompatActivity {
    private static final String DATA = "data";
    public static final int REQUEST_CODE_WORKS = 28;

    public static final String ADD = "add";
    public static final String WORK_ITEM = "workitem";
    public static final String CAR_ITEM_ID = "carId";
    public static final String REMOVE = "remove";
    public static final String EDIT = "edit";

    private WorkDataItemAdapter adapter;
    private WorkInfoDAO workDao;
    private View noWorksView;
    private CarInfoEntity car;
    private WorkInfoEntity lastAddedItem;
    private EditText searchView;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.work_list);
        setSupportActionBar(findViewById(R.id.toolbar));

        CarInfoEntity carDataItem = getIntent().getParcelableExtra(MainActivity.ITEM);
        if (carDataItem == null) {
            finish();
            return;
        }
        this.car = carDataItem;
        // DB
        DatabaseInfo databaseInfo = DatabaseInfo.Companion.init(this).getValue();
        workDao = databaseInfo.getWorkInfoDAO();

        // Recycler view and adapter
        noWorksView = findViewById(R.id.no_cars);
        noWorksView.setVisibility(View.INVISIBLE);

        adapter = new WorkDataItemAdapter(workDao.getInfo(carDataItem.getId()), this::onCheckVisibility);
        adapter.setEditWorkListener(this::onEditWork);
        searchView = findViewById(R.id.searchView);
        adapter.addFilteringBy(searchView, () -> workDao.getInfo(carDataItem.getId()));

        RecyclerView recyclerView = findViewById(R.id.recyclerView);
        recyclerView.setAdapter(adapter);
        recyclerView.setLayoutManager(new LinearLayoutManager(this, RecyclerView.VERTICAL, false));

        // add new button
        FloatingActionButton addButton = findViewById(R.id.fab);
        addButton.setOnClickListener(view -> addWork(carDataItem));

        TextView title = findViewById(R.id.workTitle);
        title.setText(String.format("%s %s %s", carDataItem.getProducer(), carDataItem.getModel(), carDataItem.getPlateNumber()));
        findViewById(R.id.backButton).setOnClickListener(v -> {
            setResult(RESULT_CANCELED);
            finish();
        });
    }

    private void addWork(CarInfoEntity dataItem) {
        Intent intent = new Intent(WorkListActivity.this, EditWorkActivity.class);
        intent.putExtra(CAR_ITEM_ID, dataItem.getId());
        startActivityForResult(intent, REQUEST_CODE_WORKS);
    }

    private void onEditWork(WorkInfoEntity dataItem, int position) {
        Intent intent = new Intent(this, EditWorkActivity.class);
        intent.putExtra(WORK_ITEM, dataItem);
        intent.putExtra(CAR_ITEM_ID, dataItem.getCarId());
        startActivityForResult(intent, REQUEST_CODE_WORKS);
    }

    public void onCheckVisibility(boolean invisible) {
        if (invisible) {
            noWorksView.setVisibility(View.INVISIBLE);
        } else {
            noWorksView.setVisibility(View.VISIBLE);
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    @Override
    protected void onResume() {
        super.onResume();
        Editable editableText = searchView.getEditableText();
        adapter.filter(editableText, lastAddedItem, workDao.getInfo(car.getId()));
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (data != null && resultCode == RESULT_OK && REQUEST_CODE_WORKS == requestCode) {
            WorkInfoEntity item = data.getParcelableExtra(WORK_ITEM);

            String command = data.getAction();
            if (ADD.equals(command)) {
                this.lastAddedItem = item;
            }
        }
    }
}

File: courses/lesson7_2_cars_rx_java/src/main/java/by/academy/lesson7/part2/WorkListActivity.kt
package by.academy.lesson7.part2

import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.view.View
import android.widget.EditText
import android.widget.TextView
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import by.academy.lesson7.part2.data.*
import com.google.android.material.floatingactionbutton.FloatingActionButton

const val REQUEST_CODE_WORKS = 28

class WorkListActivity : AppCompatActivity() {
    private lateinit var dataStorage: AbstractDataRepository
    private lateinit var workItemsAdapter: WorkDataItemAdapter2
    private lateinit var noWorksView: TextView
    private lateinit var car: CarInfoEntity
    private var lastAddedItem: WorkInfoEntity? = null
    private lateinit var searchView: EditText


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val carDataItem = intent.getParcelableExtra(CAR_ITEM) as? CarInfoEntity?
        if (carDataItem == null) {
            finish()
            return
        }
        car = carDataItem

        setContentView(R.layout.work_list)
        setSupportActionBar(findViewById(R.id.toolbar))

        // DB
        dataStorage = RepositoryFactory().getRepository(this)

        // Recycler view and adapter
        noWorksView = findViewById<TextView>(R.id.no_cars).apply { visibility = View.INVISIBLE }
        searchView = findViewById(R.id.searchView)

        workItemsAdapter = WorkDataItemAdapter2(dataStorage.getWorkInfo(carDataItem.getId()), this::onCheckVisibility).apply {
            setEditWorkListener { dataItem: WorkInfoEntity, position: Int -> onEditWork(dataItem, position) }
            addFilteringBy(searchView) { dataStorage.getWorkInfo(carDataItem.getId()) }
        }
        findViewById<RecyclerView>(R.id.recyclerView).apply {
            adapter = workItemsAdapter
            layoutManager = LinearLayoutManager(this.context, RecyclerView.VERTICAL, false)
        }

        // add new button
        val addButton = findViewById<FloatingActionButton>(R.id.fab)
        addButton.setOnClickListener { addWork(carDataItem) }

        // title
        val title = findViewById<TextView>(R.id.workTitle)
        title.text = "${carDataItem.producer} ${carDataItem.model} ${carDataItem.plateNumber}"

        //back
        findViewById<View>(R.id.backButton).setOnClickListener {
            setResult(RESULT_CANCELED)
            finish()
        }
    }

    private fun addWork(dataItem: CarInfoEntity) {
        val intent = Intent(this@WorkListActivity, EditWorkActivity::class.java)
        intent.putExtra(CAR_ITEM_ID, dataItem.getId())
        startActivityForResult(intent, REQUEST_CODE_WORKS)
    }

    private fun onEditWork(dataItem: WorkInfoEntity, position: Int) {
        val intent = Intent(this, EditWorkActivity::class.java)
        intent.putExtra(WORK_ITEM, dataItem)
        intent.putExtra(CAR_ITEM_ID, dataItem.carId)
        startActivityForResult(intent, REQUEST_CODE_WORKS)
    }

    private fun onCheckVisibility(invisible: Boolean) = if (invisible) {
        noWorksView.visibility = View.INVISIBLE
    } else {
        noWorksView.visibility = View.VISIBLE
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    override fun onResume() {
        super.onResume()
        val editableText = searchView.editableText
        workItemsAdapter.filter(editableText, lastAddedItem, dataStorage.getWorkInfo(car.getId()))
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (data != null && resultCode == RESULT_OK && REQUEST_CODE_WORKS == requestCode) {
            val command = data.action
            if (CMD_ADD == command) {
                lastAddedItem = data.getParcelableExtra(WORK_ITEM)
            }
        }
    }

}